### PR TITLE
Functional Tests - Add BO tests for Team>employees and profiles

### PR DESCRIPTION
**Pull request by nesrineabdmouleh**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Add Functional tests for BO Employees and Profiles
| Type?         | refacto
| Category?     | TE
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | no
| How to test?  | URL_FO=SHOP_URL TEST_PATH="functional/BO/catalog/advancedParameters/team/*" npm run specific-test

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/16289)

### CODE REVIEW

File: tests/puppeteer/campaigns/commonTests/loginBO.js [hunk 1]
```diff
@@ -10,4 +10,12 @@ module.exports = {
       await this.pageObjects.boBasePage.closeOnboardingModal();
     });
   },
+
+  logoutBO() {
+    it('should log out into BO', async function () {
```
Comment (boubkerbribri) [inline tests/puppeteer/campaigns/commonTests/loginBO.js]:
```suggestion
    it('should log out from BO', async function () {
```

File: tests/puppeteer/campaigns/data/faker/employee.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+const faker = require('faker');
+
+const defaultPage = ['Dashboard', 'Orders', 'Products'];
+const language = ['English (English)', 'Français (French)'];
+const permissionProfile = ['SuperAdmin', 'Logistician', 'Translator', 'Salesman'];
```
Comment (boubkerbribri) [inline tests/puppeteer/campaigns/data/faker/employee.js]: Maybe you can import it from demo data ?

File: tests/puppeteer/pages/BO/BObasePage.js [hunk 3]
```diff
@@ -15,7 +15,7 @@ module.exports = class BOBasePage extends CommonPage {
 
     // top navbar
     this.headerLogoImage = '#header_logo';
-    this.userProfileIcon = '#employee_infos';
+    this.userProfileIcon = '#employee_infos,#header_infos #header-employee-container';
```
Comment (boubkerbribri) [inline tests/puppeteer/pages/BO/BObasePage.js:18]: user profile icon selector change from a page to another ?

File: tests/puppeteer/pages/BO/advancedParameters/team/addEmployee.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    await this.clickAndWaitForNavigation(this.saveButton);
+    await this.page.waitForSelector(this.alertSuccessBlockParagraph, {visible: true});
+    return this.getTextContent(this.alertSuccessBlockParagraph);
+  }
+
+  /**
+   * Cancel page
+   * @return {Promise<void>}
+   */
+  async cancelPage() {
```
Comment (boubkerbribri) [inline tests/puppeteer/pages/BO/advancedParameters/team/addEmployee.js]: Maybe another name for the function, `cancelCreation` or just `cancel`

File: tests/puppeteer/pages/BO/advancedParameters/team/addProfile.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
+require('module-alias/register');
+const BOBasePage = require('@pages/BO/BObasePage');
+
+module.exports = class AddProfile extends BOBasePage {
+  constructor(page) {
+    super(page);
+
+    this.pageTitleCreate = 'Add new profile • prestashop';
```
Comment (boubkerbribri) [inline tests/puppeteer/pages/BO/advancedParameters/team/addProfile.js]:
```suggestion
    this.pageTitleCreate = 'Add new profile •';
```
`prestashop` depends on shopname

File: tests/puppeteer/pages/BO/advancedParameters/team/addEmployee.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
+require('module-alias/register');
+const BOBasePage = require('@pages/BO/BObasePage');
+
+module.exports = class AddPageEmployee extends BOBasePage {
+  constructor(page) {
+    super(page);
+
+    this.pageTitleCreate = 'Add new • prestashop';
```
Comment (boubkerbribri) [inline tests/puppeteer/pages/BO/advancedParameters/team/addEmployee.js]:
```suggestion
    this.pageTitleCreate = 'Add new •';
```
prestashop depends on shopname

File: tests/puppeteer/pages/BO/advancedParameters/team/employees.js [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    // List of employees
+    this.employeeGridPanel = '#employee_grid_panel';
+    this.employeeGridTitle = `${this.employeeGridPanel} h3.card-header-title`;
+    this.employeesListForm = '#employee_grid';
+    this.employeesListTableRow = `${this.employeesListForm} tbody tr:nth-child(%ROW)`;
+    this.employeesListTableColumn = `${this.employeesListTableRow} td.column-%COLUMN`;
+    this.employeesListTableToggleDropDown = `${this.employeesListTableColumn.replace('%COLUMN', 'actions')} 
+    a[data-toggle='dropdown']`;
+    this.employeesListTableDeleteLink = `${this.employeesListTableColumn.replace('%COLUMN', 'actions')} a[data-url]`;
+    this.employeesListTableEditLink = `${this.employeesListTableColumn.replace('%COLUMN', 'actions')} a[href*='edit']`;
```
Comment (boubkerbribri) [inline tests/puppeteer/pages/BO/advancedParameters/team/employees.js]:
```js
    this.employeesListTableColumn = `${this.employeesListTableRow} td.column-%COLUMN`;
    this.employeesListTableColumnAction = this.employeesListTableColumn.replace('%COLUMN', 'actions');
   
    this.employeesListTableToggleDropDown = `${this.employeesListTableColumnAction} a[data-toggle='dropdown']`;
    this.employeesListTableDeleteLink = `${this.employeesListTableColumnAction} a[data-url]`;
    this.employeesListTableEditLink = `${this.employeesListTableColumnAction} a[href*='edit']`;
```
add selector for column actions `employeesListTableColumnAction`

File: tests/puppeteer/pages/BO/advancedParameters/team/employees.js [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+        await this.selectByVisibleText(this.employeeFilterInput.replace('%FILTERBY', filterBy), value ? 'Yes' : 'No');
+        break;
+      default:
+      // Do nothing
+    }
+    // click on search
+    await Promise.all([
+      this.page.waitForNavigation({waitUntil: 'networkidle0'}),
+      this.page.click(this.filterSearchButton),
+    ]);
```
Comment (boubkerbribri) [inline tests/puppeteer/pages/BO/advancedParameters/team/employees.js]:
```js
await this.clickAndWaitForNavigation(this.filterSearchButton);
```

File: tests/puppeteer/pages/BO/advancedParameters/team/employees.js [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+  }
+
+  /**
+   * Get Value of column Displayed
+   * @param row, row in table
+   * @return {Promise<boolean|true>}
+   */
+  async getToggleColumnValue(row) {
+    if (await this.elementVisible(
+      this.employeesListColumnValidIcon.replace('%ROW', row), 100)) return true;
```
Comment (boubkerbribri) [inline tests/puppeteer/pages/BO/advancedParameters/team/employees.js]:
```suggestion
      this.employeesListColumnValidIcon.replace('%ROW', row), 
      100,
      )
   ) return true;
```

File: tests/puppeteer/pages/BO/advancedParameters/team/profiles.js [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    // List of profiles
+    this.profileGridPanel = '#profile_grid_panel';
+    this.profileGridTitle = `${this.profileGridPanel} h3.card-header-title`;
+    this.profilesListForm = '#profile_grid';
+    this.profilesListTableRow = `${this.profilesListForm} tbody tr:nth-child(%ROW)`;
+    this.profilesListTableColumn = `${this.profilesListTableRow} td.column-%COLUMN`;
+    this.profilesListTableToggleDropDown = `${this.profilesListTableColumn.replace('%COLUMN', 'actions')} 
+    a[data-toggle='dropdown']`;
+    this.profilesListTableDeleteLink = `${this.profilesListTableColumn.replace('%COLUMN', 'actions')} a[data-url]`;
+    this.profilesListTableEditLink = `${this.profilesListTableColumn.replace('%COLUMN', 'actions')} a[href*='edit']`;
```
Comment (boubkerbribri) [inline tests/puppeteer/pages/BO/advancedParameters/team/profiles.js]: Same as page employees for column actions

File: tests/puppeteer/pages/BO/advancedParameters/team/profiles.js [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+   * Go to Edit profile page
+   * @param row, row in table
+   * @return {Promise<void>}
+   */
+  async goToEditProfilePage(row) {
+    // Click on edit
+    await Promise.all([
+      this.page.click(this.profilesListTableEditLink.replace('%ROW', row)),
+      this.page.waitForNavigation({waitUntil: 'networkidle0'}),
+    ]);
```
Comment (boubkerbribri) [inline tests/puppeteer/pages/BO/advancedParameters/team/profiles.js]: replace with clickAndWaitForNavigation

File: tests/puppeteer/pages/BO/advancedParameters/team/profiles.js [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+        await this.selectByVisibleText(this.profileFilterInput.replace('%FILTERBY', filterBy), value ? 'Yes' : 'No');
+        break;
+      default:
+      // Do nothing
+    }
+    // click on search
+    await Promise.all([
+      this.page.waitForNavigation({waitUntil: 'networkidle0'}),
+      this.page.click(this.filterSearchButton),
+    ]);
```
Comment (boubkerbribri) [inline tests/puppeteer/pages/BO/advancedParameters/team/profiles.js]: same here

File: tests/puppeteer/pages/BO/BObasePage.js (hunk 3 above)
Comment (nesrineabdmouleh) [reply tests/puppeteer/pages/BO/BObasePage.js:18]: Yes the selector is changed in all pages migrated to Symphony.

File: tests/puppeteer/campaigns/data/faker/employee.js (hunk 2 above)
Comment (nesrineabdmouleh) [reply tests/puppeteer/campaigns/data/faker/employee.js]: The demo data file contains only the default employee data.
These data are in the employee creation page and they are not a demo data for an employee
You think that all data in the creation page should be added to the employee demo data ?

Comment (boubkerbribri) [reply tests/puppeteer/campaigns/data/faker/employee.js]: I'm talking about permissionProfile array, it the same as demo profiles

File: tests/puppeteer/pages/BO/advancedParameters/team/profiles.js (hunk 10 above)
Comment (nesrineabdmouleh) [reply tests/puppeteer/pages/BO/advancedParameters/team/profiles.js]: What's the problem here?

File: tests/puppeteer/campaigns/functional/BO/advancedParameters/team/01_CRUDEmployee.js [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+        const pageTitle = await this.pageObjects.addEmployeePage.getPageTitle();
+        await expect(pageTitle).to.contains(this.pageObjects.addEmployeePage.pageTitleEdit);
+      });
+
+      it('should update the employee account', async function () {
+        const textResult = await this.pageObjects.addEmployeePage.createEditEmployee(firstEditEmployeeData);
+        await expect(textResult).to.equal(this.pageObjects.addEmployeePage.successfulUpdateMessage);
+      });
+
+      it('should click on cancel and verify the new employees number', async function () {
```
Comment (PierreRambaud) [inline tests/puppeteer/campaigns/functional/BO/advancedParameters/team/01_CRUDEmployee.js]:
```suggestion
      it('should click on cancel and verify the new employee\'s number', async function () {
```

File: tests/puppeteer/pages/BO/advancedParameters/team/profiles.js (hunk 10 above)
Comment (boubkerbribri) [reply tests/puppeteer/pages/BO/advancedParameters/team/profiles.js]: it same suggestion as [here](https://github.com/PrestaShop/PrestaShop/pull/16289#discussion_r343035040)

File: tests/puppeteer/campaigns/functional/BO/advancedParameters/team/01_CRUDEmployee.js [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+      it('should sign in with edited account and verify the default page', async function () {
+        await this.pageObjects.loginPage.login(firstEditEmployeeData.email, firstEditEmployeeData.password);
+        const pageTitle = await this.pageObjects.ordersPage.getPageTitle();
+        await expect(pageTitle).to.contains(this.pageObjects.ordersPage.pageTitle);
+      });
+
+      // Logout into BO
+      loginCommon.logoutBO();
+    });
+    describe('Disable the Employee and check it', async () => {
```
Comment (PierreRambaud) [inline tests/puppeteer/campaigns/functional/BO/advancedParameters/team/01_CRUDEmployee.js]:
```suggestion
    describe('Disable the employee and check it', async () => {
```

File: tests/puppeteer/campaigns/functional/BO/advancedParameters/team/01_CRUDEmployee.js [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+
+      it('should disable the employee account', async function () {
+        const textResult = await this.pageObjects.addEmployeePage.createEditEmployee(secondEditEmployeeData);
+        await expect(textResult).to.equal(this.pageObjects.addEmployeePage.successfulUpdateMessage);
+      });
+
+      // Logout into BO
+      loginCommon.logoutBO();
+
+      it('should test sign in with the disabled Employee', async function () {
```
Comment (PierreRambaud) [inline tests/puppeteer/campaigns/functional/BO/advancedParameters/team/01_CRUDEmployee.js]:
```suggestion
      it('should test sign in with the disabled employee', async function () {
```

File: tests/puppeteer/pages/BO/advancedParameters/team/employees.js [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+    // Header links
+    this.addNewEmployeeLink = '#page-header-desc-configuration-add[title=\'Add new employee\']';
+    this.profilesTab = '#subtab-AdminProfiles';
+    // List of employees
+    this.employeeGridPanel = '#employee_grid_panel';
+    this.employeeGridTitle = `${this.employeeGridPanel} h3.card-header-title`;
+    this.employeesListForm = '#employee_grid';
+    this.employeesListTableRow = `${this.employeesListForm} tbody tr:nth-child(%ROW)`;
+    this.employeesListTableColumn = `${this.employeesListTableRow} td.column-%COLUMN`;
+    this.employeesListTableColumnAction = `${this.employeesListTableColumn.replace('%COLUMN', 'actions')}`;
```
Comment (boubkerbribri) [inline tests/puppeteer/pages/BO/advancedParameters/team/employees.js]:
```suggestion
    this.employeesListTableColumnAction = this.employeesListTableColumn.replace('%COLUMN', 'actions');
```

File: tests/puppeteer/pages/BO/advancedParameters/team/profiles.js [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+    // Selectors
+    // Header links
+    this.addNewProfileLink = '#page-header-desc-configuration-add[title=\'Add new profile\']';
+    // List of profiles
+    this.profileGridPanel = '#profile_grid_panel';
+    this.profileGridTitle = `${this.profileGridPanel} h3.card-header-title`;
+    this.profilesListForm = '#profile_grid';
+    this.profilesListTableRow = `${this.profilesListForm} tbody tr:nth-child(%ROW)`;
+    this.profilesListTableColumn = `${this.profilesListTableRow} td.column-%COLUMN`;
+    this.profilesListTableColumnAction = `${this.profilesListTableColumn.replace('%COLUMN', 'actions')}`;
```
Comment (boubkerbribri) [inline tests/puppeteer/pages/BO/advancedParameters/team/profiles.js]:
```suggestion
    this.profilesListTableColumnAction = this.profilesListTableColumn.replace('%COLUMN', 'actions');
```

File: tests/puppeteer/pages/BO/advancedParameters/team/addEmployee.js (hunk 6 above)
Comment (boubkerbribri) [reply tests/puppeteer/pages/BO/advancedParameters/team/addEmployee.js]: did you apply this modification, can I resolve conversation

File: tests/puppeteer/campaigns/data/faker/employee.js [hunk 18]
```diff
@@ -0,0 +1,18 @@
+const faker = require('faker');
+const {Profiles} = require('@data/demo/profiles');
+
+const defaultPage = ['Dashboard', 'Orders', 'Products'];
```
Comment (PierreRambaud) [inline tests/puppeteer/campaigns/data/faker/employee.js]: You should do the same for languages and defaultPage.

File: tests/puppeteer/campaigns/data/faker/employee.js [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+const language = ['English (English)', 'Français (French)'];
```
Comment (PierreRambaud) [inline tests/puppeteer/campaigns/data/faker/employee.js]:
```suggestion
const languages = ['English (English)', 'Français (French)'];
```

File: tests/puppeteer/campaigns/functional/BO/advancedParameters/team/02_CRUDEmployee.js [hunk 20]
```diff
@@ -0,0 +1,248 @@
...
+
+// Create, Read, Update and Delete Employee in BO
+describe('Create, Read, Update and Delete Employee in BO', async () => {
+  // before and after functions
+  before(async function () {
+    browser = await helper.createBrowser();
+    page = await helper.newTab(browser);
+    this.pageObjects = await init();
+    createEmployeeData = await (new EmployeeFaker({
+      defaultPage: 'Products',
```
Comment (PierreRambaud) [inline tests/puppeteer/campaigns/functional/BO/advancedParameters/team/02_CRUDEmployee.js:46]: These demo values should came from a fixtures directory? :thinking:

File: tests/puppeteer/campaigns/functional/BO/advancedParameters/team/04_CRUDProfile.js [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+
+// Create, Read, Update and Delete profile in BO
+describe('Create, Read, Update and Delete profile in BO', async () => {
+  // before and after functions
+  before(async function () {
+    browser = await helper.createBrowser();
+    page = await helper.newTab(browser);
+    this.pageObjects = await init();
+    profileData = await (new ProfileFaker({
+      defaultPage: 'Products',
```
Comment (PierreRambaud) [inline tests/puppeteer/campaigns/functional/BO/advancedParameters/team/04_CRUDProfile.js:47]: same here

File: tests/puppeteer/pages/BO/advancedParameters/team/index.js [hunk 22]
```diff
@@ -0,0 +1,219 @@
...
+  }
+
+  /**
+   * Filter list of employees
+   * @param filterType, input or select to choose method of filter
+   * @param filterBy, column to filter
+   * @param value, value to filter with
+   * @return {Promise<void>}
+   */
+  async filterEmployees(filterType, filterBy, value = '') {
```
Comment (PierreRambaud) [inline tests/puppeteer/pages/BO/advancedParameters/team/index.js:98]: This function return nothing :thinking:

File: tests/puppeteer/pages/BO/advancedParameters/team/index.js [hunk 23]
```diff
@@ -0,0 +1,219 @@
...
+    return this.elementVisible(this.employeesListColumnValidIcon.replace('%ROW', row),
+      100,
+    );
+  }
+
+  /**
+   * Quick edit toggle column value
+   * @param row, row in table
+   * @param valueWanted, Value wanted in column
+   * @return {Promise<boolean>} return true if action is done, false otherwise
```
Comment (PierreRambaud) [inline tests/puppeteer/pages/BO/advancedParameters/team/index.js:128]: this one return a boolean not a Promise :thinking:

File: tests/puppeteer/campaigns/functional/BO/advancedParameters/team/02_CRUDEmployee.js (hunk 20 above)
Comment (PierreRambaud) [reply tests/puppeteer/campaigns/functional/BO/advancedParameters/team/02_CRUDEmployee.js:46]: Seen with @boubkerbribri :+1:

**Review by PierreRambaud — APPROVED** [review]

**Pull request merged**